File: algorithms/sorting/MergeSortExample.cpp
#include <iostream>
#include <vector>

// Function for merging two sorted segments
void merge(std::vector<int>& arr, int start, int mid, int end) {
    int n1 = mid - start + 1;
    int n2 = end - mid;

    // Create temp arrays
    std::vector<int> L(n1);
    std::vector<int> R(n2);

    // Copy data to temp arrays L[] and R[]
    for (int i = 0; i < n1; i++)
        L[i] = arr[start + i];
    for (int j = 0; j < n2; j++)
        R[j] = arr[mid + 1 + j];

    // Merge the temp arrays back into arr[start..end]
    int i = 0;
    int j = 0;
    int k = start;

    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            arr[k] = L[i];
            i++;
        } else {
            arr[k] = R[j];
            j++;
        }
        k++;
    }

    // Copy the remaining elements of L[], if there are any
    while (i < n1) {
        arr[k] = L[i];
        i++;
        k++;
    }

    // Copy the remaining elements of R[], if there are any
    while (j < n2) {
        arr[k] = R[j];
        j++;
        k++;
    }
}

// start is for left index and end is right index of the sub-array of arr to be sorted
void mergeSort(std::vector<int>& arr, int start, int end) {
    if (start >= end)
        return; // Returns recursively

    int mid = start + (end - start) / 2;
    mergeSort(arr, start, mid); // Sorting first half
    mergeSort(arr, mid + 1, end); // Sorting second half
    merge(arr, start, mid, end); // Merging the sorted halves
}

int main() {
    // Our main function
    std::vector<int> arr = {12, 11, 13, 5, 6, 7};
    int arr_size = arr.size();

    std::cout << "Given array is \n";
    for (auto i : arr)
        std::cout << i << " ";

    mergeSort(arr, 0, arr_size - 1);

    std::cout << "\nSorted array is \n";
    for (auto i : arr)
        std::cout << i << " ";
    std::cout << std::endl;
    return 0;
}

File: algorithms/sorting/MergeSortExample.java
import java.util.Arrays;

public class MergeSortExample {

    // Method to merge two sorted subarrays
    public static void merge(int[] arr, int left, int middle, int right) {
        // Find sizes of two subarrays to be merged
        int n1 = middle - left + 1;
        int n2 = right - middle;

        // Create temp arrays
        int[] leftArray = new int[n1];
        int[] rightArray = new int[n2];

        // Copy data to temp arrays
        System.arraycopy(arr, left, leftArray, 0, n1);
        for (int j = 0; j < n2; ++j)
            rightArray[j] = arr[middle + 1 + j];

        // Merge the temp arrays

        // Initial indexes of first and second subarrays
        int i = 0, j = 0;

        // Initial index of merged subarray array
        int k = left;
        while (i < n1 && j < n2) {
            if (leftArray[i] <= rightArray[j]) {
                arr[k] = leftArray[i];
                i++;
            } else {
                arr[k] = rightArray[j];
                j++;
            }
            k++;
        }

        // Copy remaining elements of leftArray[] if any
        while (i < n1) {
            arr[k] = leftArray[i];
            i++;
            k++;
        }

        // Copy remaining elements of rightArray[] if any
        while (j < n2) {
            arr[k] = rightArray[j];
            j++;
            k++;
        }
    }

    // Method to sort the array using the merge sort algorithm
    public static void sort(int[] arr, int left, int right) {
        if (left < right) {
            // Find the middle point
            int middle = left + (right - left) / 2;

            // Sort first and second halves
            sort(arr, left, middle);
            sort(arr, middle + 1, right);

            // Merge the sorted halves
            merge(arr, left, middle, right);
        }
    }

    public static void main(String[] args) {
        // Test the merge sort algorithm
        int[] actual = { 5, 1, 6, 2, 3, 4 };
        System.out.println("Original array: " + Arrays.toString(actual));

        sort(actual, 0, actual.length - 1);
        System.out.println("Sorted array: " + Arrays.toString(actual));
    }
}
